File: haskell/sieve/haskell/sieve/Sieve.hs
module Sieve (primesUpTo) where

primesUpTo :: Int -> [Int]
primesUpTo limit = filter (<= limit) (2 : sieve [3, 5..limit])

sieve :: [Int] -> [Int]	
sieve [] = []
sieve (p:xs) = p : sieve [x | x <- xs, x `mod` p > 0]

File: prolog/sieve/sieve.pl
:- use_module(library(clpfd)).

primes(Limit, []) :- Limit #< 2, !.
primes(Limit, Primes) :-
    numlist(2, Limit, Sieve),
    Stop is round(sqrt(Limit)),
    primes(Stop, Sieve, Primes).

primes(Stop, [N|Sieve], [N|Primes]):-
  exclude(is_multiple(N), Sieve, Sieved),
  (N #=< Stop -> primes(Stop, Sieved, Primes) ; Primes=Sieved).
  
is_multiple(X, Y) :- Y mod X #= 0.
